Implement WebTransport from a Worker

achristensen07 opened the pull request:
#### c85047d65d8d007e687fb0ab36e069536cc1028e
<pre>
Implement WebTransport from a Worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=284965">https://bugs.webkit.org/show_bug.cgi?id=284965</a>
<a href="https://rdar.apple.com/136263473">rdar://136263473</a>

Reviewed by Youenn Fablet.

A Worker runs code off the main thread, so we need to be extra careful.

The first thing I noticed is that ensureNetworkProcessConnection can only
safely be called from the main thread, so if we are creating a WebTransport
session off the main thread, hop to the main thread, get the connection,
then hop back.  The connection can then be safely used off the main thread.

To make networking fast, especially the sending of datagrams, I send IPC
from the background thread to the network process, which is already safe.
Some of the messages need a reply, and some messages indicate data or streams
were received in the network process, and those messages come from the
network process on the main thread.  To handle the needed thread hopping,
I introduce WorkerWebTransportSession which works as an intermediary between
the WebTransportSession and the WebTransportSessionClient to make sure
everything is safely on the expected thread on both sides.

To make use of m_readStreamSources safe, I needed to change the lookup in
WebTransportSession::streamReceiveBytes into just a client call and have
the WorkerWebTransportSession take care of the thread safety.  That means
I needed to move WebTransportStreamIdentifier to WebCore and have WebCore
keep track of the map of identifier to stream and only access that map on
the proper thread.

I found that I can use WebCore::WebTransportReceiveStreamSource to receive both
data and streams, so I removed WebKit::WebTransportReceiveStreamSource.

* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::initializeOverHTTP):
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::createBidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::streamReceiveBytes):
(WebCore::WebTransport::cleanup):
(WebCore::WebTransport::createBidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.cpp:
(WebCore::WebTransportBidirectionalStreamConstructionParameters::WebTransportBidirectionalStreamConstructionParameters):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveBytes):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp: Added.
(WebCore::WorkerWebTransportSession::create):
(WebCore::WorkerWebTransportSession::WorkerWebTransportSession):
(WebCore::WorkerWebTransportSession::receiveDatagram):
(WebCore::WorkerWebTransportSession::receiveIncomingUnidirectionalStream):
(WebCore::WorkerWebTransportSession::receiveBidirectionalStream):
(WebCore::WorkerWebTransportSession::streamReceiveBytes):
(WebCore::WorkerWebTransportSession::sendDatagram):
(WebCore::WorkerWebTransportSession::createOutgoingUnidirectionalStream):
(WebCore::WorkerWebTransportSession::createBidirectionalStream):
(WebCore::WorkerWebTransportSession::terminate):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h: Added.
(WebCore::WorkerWebTransportSession::ref const):
(WebCore::WorkerWebTransportSession::deref const):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::sendStreamSendBytes):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::destroyOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::destroyBidirectionalStream):
(WebKit::NetworkTransportSession::streamReceiveBytes):
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.cpp:
(WebKit::NetworkTransportStream::NetworkTransportStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
(WebKit::NetworkTransportStream::identifier const):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::createParameters):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createStream):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::NetworkTransportStream):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::dispatchMessage):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp: Removed.
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.h: Removed.
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
(WebKit::WebTransportSendStreamSink::WebTransportSendStreamSink):
(WebKit::WebTransportSendStreamSink::write):
(WebKit::WebTransportSendStreamSink::sendBytes):
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.h:
(WebKit::WebTransportSendStreamSink::create):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
(WebKit::WebTransportSession::WebTransportSession):
(WebKit::WebTransportSession::messageSenderConnection const):
(WebKit::WebTransportSession::messageSenderDestinationID const):
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::streamReceiveBytes):
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
(WebKit::WebTransportSession::streamSendBytes):
(WebKit::WebTransportSession::terminate):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in:
* Source/WebKit/WebProcess/Network/WorkerWebTransportReceiveStreamSource.cpp: Added.
* Source/WebKit/WebProcess/Network/WorkerWebTransportReceiveStreamSource.h: Added.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::webTransportSession):
(WebKit::WebProcess::addWebTransportSession):
(WebKit::WebProcess::removeWebTransportSession):
* Source/WebKit/WebProcess/WebProcess.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_Worker)):

Canonical link: <a href="https://commits.webkit.org/288367@main">https://commits.webkit.org/288367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f9eb7272921ede6b7e7fa1df775f5c0cae6c7b6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82877 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37151 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/87995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/33932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10442 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/87995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/33932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85933 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/1931 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75389 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/87995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/82314 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29579 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32966 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/30257 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89361 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10168 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7394 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10396 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71195 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/72221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16385 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12819 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10121 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9993 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13458 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11761 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
```diff
@@ -0,0 +1,150 @@
+    , m_client(client)
+    , m_session(WTFMove(session))
+{
+    m_session->attachClient(*this);
```
I don't understand how safer cpp thinks this is uncounted and unsafe.  m_session is a Ref

cdumez replied:
If `m_session` never gets reassigned then mark it as `const`. Otherwise, you need a `Ref` the the session *on the stack* for safety.

rniwa commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
```diff
@@ -0,0 +1,160 @@
+    , m_client(client)
+    , m_session(WTFMove(session))
+{
+    protectedSession()->attachClient(*this);
```
You don't need to use protectedSession() if you've made m_session const.

youennf commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -45,18 +46,31 @@ RefPtr<ThreadableWebSocketChannel> WebSocketProvider::createWebSocketChannel(Doc
+            WebKit::WebTransportSession::initialize(WebProcess::singleton().ensureNetworkProcessConnection().connection(), url, webPageProxyID, origin, [context = WTFMove(context), completionHandler = WTFMove(completionHandler)] (RefPtr<WebTransportSession>&& session) mutable {
+                ASSERT(RunLoop::isMain());
+                auto& scriptExecutionContext = context.get();
+                scriptExecutionContext.postTask([completionHandler = WTFMove(completionHandler), session = WTFMove(session), context = WTFMove(context)] (auto&) mutable {
```
I don't think it is safe to ref the context since postTask may fail, leading to context being derefed in main thread.
Instead, let's pass a `ScriptExecutionContextIdentifier` and use the `static postTaskTo` version.
Ditto for completionHandler which could, in case of stopping a worker:
1. never be called
2. be destroyed in main thread
Let's register the completion handler in a map on WorkerGlobalScope or something like that.

Given this is using a completion handler, this is likely using an asynchronous reply.
In such a case, the combo NativePromise IPC + enqueueTaskWhenSettled is probably the best approach to use completion handler without worrying about this kind of issues.

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -110,13 +112,18 @@ void WebTransport::initializeOverHTTP(SocketProvider& provider, ScriptExecutionC
-    provider.initializeWebTransportSession(context, url, [this, protectedThis = Ref { *this }] (RefPtr<WebTransportSession>&& session) {
-        m_session = WTFMove(session);
-        if (!m_session) {
+    provider.initializeWebTransportSession(context, url, [this, protectedThis = Ref { *this }, scope = RefPtr { dynamicDowncast<WorkerGlobalScope>(context) }] (RefPtr<WebTransportSession>&& session) {
```
Can this completion handler be destroyed in main thread in case of a worker being stopped during the async reply?

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -156,24 +163,27 @@ void WebTransport::receiveDatagram(std::span<const uint8_t> datagram)
     m_receiveStreams.add(receiveStream);
-    m_receiveStreamSource->receiveIncomingStream(jsDOMGlobalObject, WTFMove(receiveStream));
+    ASSERT(!m_readStreamSources.contains(identifier));
+    m_readStreamSources.set(identifier, WTFMove(incomingStream));
```
```suggestion
    m_readStreamSources.add(identifier, WTFMove(incomingStream));
```
Given the map cannot contain identifier?

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -156,24 +163,27 @@ void WebTransport::receiveDatagram(std::span<const uint8_t> datagram)
-    m_receiveStreamSource->receiveIncomingStream(jsDOMGlobalObject, WTFMove(receiveStream));
+    ASSERT(!m_readStreamSources.contains(identifier));
+    m_readStreamSources.set(identifier, WTFMove(incomingStream));
+    m_receiveStreamSource->receiveIncomingStream(jsDOMGlobalObject, stream.releaseReturnValue());
```
We are calling twice `releaseReturnValue`, should it be WTFMove(receiveStream) instead?

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -199,7 +209,10 @@ void WebTransport::receiveBidirectionalStream(WebTransportBidirectionalStreamCon
-    auto stream = WebCore::createBidirectionalStream(jsDOMGlobalObject, WTFMove(parameters));
+    auto incomingStream = WebTransportReceiveStreamSource::create();
+    ASSERT(!m_readStreamSources.get(parameters.identifier));
+    m_readStreamSources.set(parameters.identifier, incomingStream.copyRef());
```
add?

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -208,6 +221,14 @@ void WebTransport::receiveBidirectionalStream(WebTransportBidirectionalStreamCon
+    if (RefPtr source = m_readStreamSources.get(identifier))
+        source->receiveBytes(span, withFin);
+    else
+        ASSERT_NOT_REACHED();
```
I would write it ASSERT(m_readStreamSources.contains(identifier))` instead.

youennf commented on Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
```diff
@@ -46,4 +46,15 @@ void WebTransportReceiveStreamSource::receiveIncomingStream(JSC::JSGlobalObject&
+{
+    if (m_isCancelled)
+        return;
+    auto arrayBuffer = ArrayBuffer::tryCreateUninitialized(bytes.size(), 1);
```
How about `ArrayBuffer::tryCreate(std::span<const uint8_t>)`?

youennf commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
```diff
@@ -0,0 +1,160 @@
+void WorkerWebTransportSession::receiveDatagram(std::span<const uint8_t> span)
+{
+    ASSERT(RunLoop::isMain());
+    protectedScope()->protectedThread()->runLoop().postTaskForMode([vector = Vector<uint8_t> { span }, weakClient = m_client] (auto&) mutable {
```
I do not think it is safe to ref scope in main thread given it is RefCounted.
WorkerThread can probably be refer, or just use ScriptExecutionContext::postTaskTo.

youennf commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
```diff
@@ -0,0 +1,66 @@
+class WebTransport;
+class WorkerGlobalScope;
+
+class WorkerWebTransportSession : public WebTransportSessionClient, public WebTransportSession {
```
Does it need to be `public WebTransportSessionClient` or can it be `private WebTransportSessionClient`?

youennf commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
```diff
@@ -0,0 +1,66 @@
+
+    const Ref<WorkerGlobalScope> m_scope;
+    ThreadSafeWeakPtr<WebTransportSessionClient> m_client;
+    const Ref<WebTransportSession> m_session;
```
I would rename this one to make it clear it is the main thread session.
Note that this m_session needs to be able to be derefed from non main thread.

achristensen07 replied:
I think I disagree with this comment.  This session is not a "main thread" session, and all WebTransportSessions are safe to ref and defer from any thread.

youennf commented on Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
```diff
@@ -50,10 +52,8 @@ enum class NetworkTransportStreamType : uint8_t;
 using WebTransportSessionIdentifier = ObjectIdentifier<WebTransportSessionIdentifierType>;
-using WebTransportStreamIdentifier = ObjectIdentifier<WebTransportStreamIdentifierType>;
 
 class NetworkTransportSession : public RefCounted<NetworkTransportSession>, public IPC::MessageReceiver, public IPC::MessageSender, public Identified<WebTransportSessionIdentifier> {
```
Ideally, we would make it WorkQueue message receiver so that all processing is done out of the main thread in network process from the start.

achristensen07 replied:
This is more involved.  We'd have to pipe that into all the existing calls of dispatch_get_main_queue in NetworkTransport* files.  I think this should be a separate PR.

youennf commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -45,18 +46,31 @@ RefPtr<ThreadableWebSocketChannel> WebSocketProvider::createWebSocketChannel(Doc
     return WebKit::WebSocketChannel::create(m_webPageProxyID, document, client);
 }
 
-void WebSocketProvider::initializeWebTransportSession(WebCore::ScriptExecutionContext& context, const URL& url, CompletionHandler<void(RefPtr<WebCore::WebTransportSession>&&)>&& completionHandler)
```
We should probably rename `WebSocketProvider` to make it a bit more abstract as a follow-up.

achristensen07 replied:
There is a FIXME comment to do this already in WebTransport::initializeOverHTTP

youennf commented on Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
```diff
@@ -64,20 +65,33 @@ void WebTransportSendStreamSink::write(WebCore::ScriptExecutionContext& context,
+        });
+    }, CompletionHandlerCallThread::AnyThread };
+
+    session->streamSendBytes(m_identifier, bytes, false, WTFMove(callCompletionHandlerOnScriptThread));
```
I would tend to use NativePromise/enqueueTaskWhenSettled.

youennf commented on Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
```diff
@@ -64,20 +65,33 @@ void WebTransportSendStreamSink::write(WebCore::ScriptExecutionContext& context,
+            return completionHandler();
+        scriptExecutionContext.postTask([
+            context = WTFMove(context),
+            completionHandler = WTFMove(completionHandler)
```
Ditto here, completion handler may be destroyed in main thread if postTask fails.

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -199,7 +210,10 @@ void WebTransport::receiveBidirectionalStream(WebTransportBidirectionalStreamCon
     auto& jsDOMGlobalObject = *JSC::jsCast<JSDOMGlobalObject*>(globalObject);
-    auto stream = WebCore::createBidirectionalStream(jsDOMGlobalObject, WTFMove(parameters));
+    auto incomingStream = WebTransportReceiveStreamSource::create();
+    ASSERT(!m_readStreamSources.get(parameters.identifier));
```
```suggestion
    ASSERT(!m_readStreamSources.contains(parameters.identifier));
```

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -156,24 +164,27 @@ void WebTransport::receiveDatagram(std::span<const uint8_t> datagram)
     if (!globalObject)
         return;
     auto& jsDOMGlobalObject = *JSC::jsCast<JSDOMGlobalObject*>(globalObject);
+    auto incomingStream = WebTransportReceiveStreamSource::create();
```
It seems we now prefer Ref over auto, here and probably elsewhere.

youennf commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
```diff
@@ -0,0 +1,63 @@
+
+class WebTransport;
+
+class WorkerWebTransportSession : public WebTransportSession, private WebTransportSessionClient {
```
We can keep it as is but I wonder whether WorkerWebTransportSession could not be merged with WebCore:: WebTransportSession provided WebTransportSession would know the context ID of its client.

youennf commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -45,22 +46,35 @@ RefPtr<ThreadableWebSocketChannel> WebSocketProvider::createWebSocketChannel(Doc
+            url = crossThreadCopy(url)
+        ] mutable {
+            WebKit::WebTransportSession::initialize(WebProcess::singleton().ensureNetworkProcessConnection().connection(), url, webPageProxyID, origin)->whenSettled(RunLoop::protectedMain(), [producer = WTFMove(producer), contextID] (auto&& result) mutable {
+                ScriptExecutionContext::postTaskTo(contextID, [producer = WTFMove(producer), result = WTFMove(result)] (auto&) mutable {
```
I am not sure this is needed, resolving the producer does not need to be done on the context thread.

youennf commented on Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
```diff
@@ -48,8 +48,8 @@ WebTransportSendStreamSink::~WebTransportSendStreamSink()
 
 void WebTransportSendStreamSink::write(WebCore::ScriptExecutionContext& context, JSC::JSValue value, WebCore::DOMPromiseDeferred<void>&& promise)
```
I would be tempted (in a follow-up) to move back the DOMPromise code in WebCore.

youennf reviewed (approved):
LGTM

achristensen07 committed:
Implement WebTransport from a Worker
https://bugs.webkit.org/show_bug.cgi?id=284965
rdar://136263473

Reviewed by Youenn Fablet.

A Worker runs code off the main thread, so we need to be extra careful.

The first thing I noticed is that ensureNetworkProcessConnection can only
safely be called from the main thread, so if we are creating a WebTransport
session off the main thread, hop to the main thread, get the connection,
then hop back.  The connection can then be safely used off the main thread.

To make networking fast, especially the sending of datagrams, I send IPC
from the background thread to the network process, which is already safe.
Some of the messages need a reply, and some messages indicate data or streams
were received in the network process, and those messages come from the
network process on the main thread.  To handle the needed thread hopping,
I introduce WorkerWebTransportSession which works as an intermediary between
the WebTransportSession and the WebTransportSessionClient to make sure
everything is safely on the expected thread on both sides.

To make use of m_readStreamSources safe, I needed to change the lookup in
WebTransportSession::streamReceiveBytes into just a client call and have
the WorkerWebTransportSession take care of the thread safety.  That means
I needed to move WebTransportStreamIdentifier to WebCore and have WebCore
keep track of the map of identifier to stream and only access that map on
the proper thread.

I found that I can use WebCore::WebTransportReceiveStreamSource to receive both
data and streams, so I removed WebKit::WebTransportReceiveStreamSource.

* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::initializeOverHTTP):
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::createBidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::streamReceiveBytes):
(WebCore::WebTransport::cleanup):
(WebCore::WebTransport::createBidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.cpp:
(WebCore::WebTransportBidirectionalStreamConstructionParameters::WebTransportBidirectionalStreamConstructionParameters):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveBytes):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp: Added.
(WebCore::WorkerWebTransportSession::create):
(WebCore::WorkerWebTransportSession::WorkerWebTransportSession):
(WebCore::WorkerWebTransportSession::receiveDatagram):
(WebCore::WorkerWebTransportSession::receiveIncomingUnidirectionalStream):
(WebCore::WorkerWebTransportSession::receiveBidirectionalStream):
(WebCore::WorkerWebTransportSession::streamReceiveBytes):
(WebCore::WorkerWebTransportSession::sendDatagram):
(WebCore::WorkerWebTransportSession::createOutgoingUnidirectionalStream):
(WebCore::WorkerWebTransportSession::createBidirectionalStream):
(WebCore::WorkerWebTransportSession::terminate):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h: Added.
(WebCore::WorkerWebTransportSession::ref const):
(WebCore::WorkerWebTransportSession::deref const):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::sendStreamSendBytes):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::destroyOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::destroyBidirectionalStream):
(WebKit::NetworkTransportSession::streamReceiveBytes):
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.cpp:
(WebKit::NetworkTransportStream::NetworkTransportStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
(WebKit::NetworkTransportStream::identifier const):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::createParameters):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createStream):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::NetworkTransportStream):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::dispatchMessage):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp: Removed.
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.h: Removed.
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
(WebKit::WebTransportSendStreamSink::WebTransportSendStreamSink):
(WebKit::WebTransportSendStreamSink::write):
(WebKit::WebTransportSendStreamSink::sendBytes):
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.h:
(WebKit::WebTransportSendStreamSink::create):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
(WebKit::WebTransportSession::WebTransportSession):
(WebKit::WebTransportSession::messageSenderConnection const):
(WebKit::WebTransportSession::messageSenderDestinationID const):
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::streamReceiveBytes):
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
(WebKit::WebTransportSession::streamSendBytes):
(WebKit::WebTransportSession::terminate):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in:
* Source/WebKit/WebProcess/Network/WorkerWebTransportReceiveStreamSource.cpp: Added.
* Source/WebKit/WebProcess/Network/WorkerWebTransportReceiveStreamSource.h: Added.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::webTransportSession):
(WebKit::WebProcess::addWebTransportSession):
(WebKit::WebProcess::removeWebTransportSession):
* Source/WebKit/WebProcess/WebProcess.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_Worker)):

Canonical link: https://commits.webkit.org/288367@main
